app/hub/Views/UpgradePlanView/UpgradePlanView.jsx

```diff
@@ -0,0 +1,957 @@
+											<span className="x-icon" />
+										</td>
+										<td className="col-premium">
+											<span className="check purple" />
+										</td>
+									</tr>
+									<tr className="gray">
+										<td colSpan="3">
+											<span className="feature-title">{t('hub_upgrade_unlimited_bandwidth')}</span>
+											<span className="premium-sparkle" />
+										</td>
+									</tr>
+									<tr className="gray border-bottom">
+										<td>
+											<span className="x-icon" />
+										</td>
+										<td>
+											<span className="x-icon" />
+										</td>
+										<td className="col-premium">
+											<span className="check purple" />
+										</td>
+									</tr>
+								</tbody>
+							</table>
+						</div>
+					</div>
+					<div className="row align-center footer-buttons">
+						<div className="small-12 text-center columns">
+							<span className="col-free">
+								<NavLink className="button already-protected" to="/home" title="Already Protected">
+									{t('hub_upgrade_already_protected')}
+								</NavLink>
+							</span>
+						</div>
+						<div className="small-12 text-center columns">
+							<span className="col-plus">
+								{isPlus ? (
+									<NavLink className="button button-gold" to="/home" title="Already Protected">
+										{t('hub_upgrade_already_protected')}
+									</NavLink>
+								) : (
+									<a className="button button-gold" href={plusCheckoutLink} target="_blank" rel="noopener noreferrer" title="Upgrade to Plus">
+										{t('hub_upgrade_to_plus')}
+									</a>
+								)}
+							</span>
+						</div>
+						<div className="small-12 text-center columns">
+							<span className="col-premium">
+								{isPremium ? (
+									<NavLink className="button button-premium" to="/home" title="Already Protected">
+										{t('hub_upgrade_already_protected')}
+									</NavLink>
+								) : (
+									<a className="button button-premium" href={premiumCheckoutLink} target="_blank" rel="noopener noreferrer" title="Upgrade to Premium">
+										{t('hub_upgrade_to_premium')}
+									</a>
+								)}
+							</span>
+						</div>
+					</div>
+				</div>
+			</div>
+		</section>
+	);
+};
+
+// PropTypes ensure we pass required props of the correct type
+UpgradePlanView.propTypes = {
+	protection_level: PropTypes.string.isRequired,
+	show_yearly_prices: PropTypes.bool.isRequired,
+	user: PropTypes.shape({
+		email: PropTypes.string,
+		plusAccess: PropTypes.bool,
+		premiumAccess: PropTypes.bool,
+	}),
+	actions: PropTypes.shape({
+		toggleMonthlyYearlyPrices: PropTypes.func.isRequired,
+		setBasicProtection: PropTypes.func.isRequired,
+		setPlusProtection: PropTypes.func.isRequired,
+		setPremiumProtection: PropTypes.func.isRequired,
+	}).isRequired,
+};
+
+// Default props used on the Home View
+UpgradePlanView.defaultProps = {
+	user: {
+		email: '',
+		plusAccess: false,
+		premiumAccess: false,
+	},
+};
+
+export default UpgradePlanView;
```

Comment:
There's a lot of duplication in the `return ( ....)` code, which makes the code harder to read and modify than it needs to be and increases bug surface area. As the last push on this PR, let's DRY it out. As one example, instead of inlining all the feature matrix rows, we can define (at file scope) a render helper with the following signature:

```
const featureMatrixRow = (label, inBasic, inPlus, isSparkle) => (
    <tr>
        ...
    </tr>
);
```

This makes it possible to condense the ~150 lines of inline `<tr>` elements to:

```
{featureMatrixRow(t('hub_upgrade_browser_tracker_blocking'), true, true)}
{featureMatrixRow(t('hub_upgrade_browser_ad_blocking'), true, true)}
{featureMatrixRow(t('hub_upgrade_custom_blocking_preferences'), true, true)}
{featureMatrixRow(t('hub_upgrade_extension_themes'), false, true)}
{featureMatrixRow(t('hub_upgrade_historical_extension_stats'), false, true)}
{featureMatrixRow(t('hub_upgrade_application_tracker_blocking'), false, true, true)}
{featureMatrixRow(t('hub_upgrade_application_ad_blocking'), false, true, true)}
{featureMatrixRow('VPN', false, false, true)}
{featureMatrixRow(t('hub_upgrade_no_vpn_logs'), false, false, true)}
{featureMatrixRow(`P2P ${t('support')}`, false, false, true)}
{featureMatrixRow(`IPv6 ${t('hub_upgrade_leak_protection')}`, false, false, true)}
{featureMatrixRow(t('hub_upgrade_physical_servers'), false, false, true)}
{featureMatrixRow(t('hub_upgrade_unlimited_bandwidth'), false, false, true)}
```

This is one of the main DRYing opportunities, but there are other big ones. At the least, we should do the same for the same batch of `<tr>` elements for the mobile view. See if you can identify two or three other spots though! We don't need to get this code perfectly DRY, but we should wring it out at least SOME so it doesn't get that mildew smell.

Reply:
Yes, we can definitely do this! I wound up DRYing the matrix rows, cards, CTA buttons, and monthly/yearly toggle and it looks much cleaner now. I declared a few of those in component scope instead of file scope so they would have access to props--let me know if there's a better way to do this. As side note, I like how you're intentionally not passing params to those render helper functions so they evaluate to falsy, I used the same trick in a few of the other functions

Reply:
Great! This is much cleaner, thank you.
Re: the ones you declared in component scope, with how many props / args we would need to pass to them, I think it's fine to leave them in there. If we get any complaints about performance during QA we know what to try, but I would be very surprised if we did.